### PR TITLE
feat: advertise hami-core Devcore via deviceCoreScaling

### PR DESCRIPTION
**Pull request by curry30000**
Scheduler counterpart: https://github.com/Project-HAMi/HAMi/pull/2952
That HAMi PR keeps a 100-point hami-core Fit budget unless this plugin advertises Devcore > 100.

**What this PR does / why we need it**:

NVIDIA already applies `deviceCoreScaling` when the plugin registers `Devcore`. This plugin hardcoded `Devcore = 100` in hami-core (`HamiVnpuCoreMaxPercent`), so compute oversell could not be advertised no matter what the scheduler did.

- `vnpus.deviceCoreScaling` (default 1), optional per-node override in `NodeConfig`
- chart value `hamiVnpuCore.deviceCoreScaling`, schema `minimum: 1`
- hami-core registers `Devcore = round(100 * deviceCoreScaling)`
- template / hard-slice still reports hardware `AICore`
- default 1 keeps today's 100-point budget

The scheduler keeps a 100-point Fit budget unless advertised `Devcore > 100`, so this plugin change is what actually enables oversell. Physical AICore counts (8/20/24/30) are not treated as that percentage.

The first commit left `deviceCoreScaling: 1` as a literal inside `deviceConfig` (`# @ignored`), so `--set` / a values file could not reach it. `c98d025` renders `{{ .Values.hamiVnpuCore.deviceCoreScaling }}` next to `hamiVnpuCore.enabled`. Fractional ratios need `--set-json` (`--set 1.5` becomes a string and the schema rejects the type). That is in the chart README.

```console
$ helm template adp ./charts/ascend-device-plugin --set hamiVnpuCore.enabled=true \
    --set-json hamiVnpuCore.deviceCoreScaling=1.5 | grep deviceCoreScaling
      deviceCoreScaling: 1.5

$ helm template adp ./charts/ascend-device-plugin --set-json hamiVnpuCore.deviceCoreScaling=0.5
Error: values don't meet the specifications of the schema(s) in the following chart(s):
ascend-device-plugin:
- at '/hamiVnpuCore/deviceCoreScaling': minimum: got 0.5, want 1
```

Ratios below 1 are rejected. `AdvertisedDevcore(true, 0.5, 20)` used to return 50, but HAMi cannot tell that apart from a hardware AICore count (8/20/24/30) and clamps anything `<= 100` back to the base, so the cut disappeared with no log. The plugin now falls back to the percentage base and warns; NaN / Inf / out-of-int32 warn too. Rounding: `1.234` -> `123`. Per-node override is shown in `ascend-device-node-configmap.yaml` and the chart README.

`f53444f` only corrects the `HamiVnpuCoreMaxPercent` comment. `registerHAMi` now calls `internal.AdvertisedDevcore`, so 100 is the base percentage, not the advertised total once scaling is above 1. No behavior change.

**Special notes for your reviewer**:

AI assistance: I used Cursor to draft the patch. I reviewed the registration path. Commit messages were written by me.

Validation: `go test ./internal/...` passes, `helm lint` is clean, `README.md` / `values.schema.json` regenerated with `make update-chart-docs` so `make verify-helm-chart` stays green.

## Summary by CodeRabbit

* **New Features**
  * Added configurable VNPU device core scaling with global and per-node overrides.
  * Supports fractional scaling ratios for compute overselling.
  * Advertised compute capacity now reflects scaling while memory and virtual-device limits remain unchanged.
  * Invalid, unsupported, or below-minimum values fall back safely to the default behavior.

* **Documentation**
  * Added Helm configuration guidance, validation details, examples, and compute overselling limitations.

* **Tests**
  * Added coverage for defaults, overrides, rounding, and invalid values.

### CODE REVIEW

**Comment by curry30000** [comment]
HAMi #2902 was closed because Fit treated `Totalcore` as a percentage. The follow-up is https://github.com/Project-HAMi/HAMi/pull/2947: keep the 100-point budget for physical AICore, and honor `Devcore > 100` advertised by this plugin (`deviceCoreScaling`).

**Comment by curry30000** [comment]
Pushed a follow-up commit (`c98d025`) fixing two problems found while re-reviewing this together with the HAMi side (Project-HAMi/HAMi#2947).

### `deviceCoreScaling` was not reachable from Helm values

`values.yaml` set it as a literal inside the `deviceConfig` template:

```yaml
deviceConfig: |- # @schema required:true
  vnpus:
    hamiVnpuCore: {{ .Values.hamiVnpuCore.enabled }}
    deviceCoreScaling: 1
```

`deviceConfig` is marked `# @ignored`, so neither `--set` nor a values file could reach the ratio. The only way to change it was to replace the entire `deviceConfig` blob, chip table included. The knob was effectively not exposed at all.

It is now a real value next to `hamiVnpuCore.enabled`, rendered into the template, with `minimum: 1` in the generated schema:

```console
$ helm template adp ./charts/ascend-device-plugin --set hamiVnpuCore.enabled=true \
    --set-json hamiVnpuCore.deviceCoreScaling=1.5 | grep deviceCoreScaling
      deviceCoreScaling: 1.5

$ helm template adp ./charts/ascend-device-plugin --set-json hamiVnpuCore.deviceCoreScaling=0.5
Error: values don't meet the specifications of the schema(s) in the following chart(s):
ascend-device-plugin:
- at '/hamiVnpuCore/deviceCoreScaling': minimum: got 0.5, want 1
```

`--set-json` is needed for fractional ratios because plain `--set` parses `1.5` as a string, which the schema then rejects on type. That is called out in the chart README.

### Ratios below 1 were silently ignored

`AdvertisedDevcore(true, 0.5, 20)` returned 50, but HAMi cannot distinguish an under-provisioned percentage from a hardware AICore count — 8/20/24/30 all land in the same range. Its `hamiCoreFitBudget` therefore clamps anything at or below 100 back to the base, and the reduced budget just disappears. The user got no error and no log.

Since it is not representable end to end, it is now rejected explicitly: fall back to the percentage base and warn. The chart schema blocks it earlier. NaN, Inf and out-of-int32 values warn now too instead of falling through silently.

### Also

- `hamiCorePercentBase` replaces the literal `100`.
- The per-node `deviceCoreScaling` override is shown in `ascend-device-node-configmap.yaml` and the chart README. `NodeConfig` already supported it but nothing documented it.
- Test cases for the rejected ratios and for rounding (`1.234` gives 123).

### Validation

`go test ./internal/` passes, `helm lint` is clean, and `README.md` / `values.schema.json` are regenerated with `make update-chart-docs` so `make verify-helm-chart` stays green.

**Comment by curry30000** [comment]
Pushed a third commit fixing a comment this PR made inaccurate.

`registerHAMi` no longer uses `HamiVnpuCoreMaxPercent`, since it now calls `internal.AdvertisedDevcore`. The constant's doc comment still described 100 as "the total allocatable core units per device", which stops being true as soon as `deviceCoreScaling` is above 1 — the advertised total becomes `round(100 * deviceCoreScaling)`. The comment now describes it as the base percentage and points at `AdvertisedDevcore`.

Comment only, no behavior change. `go test ./internal/` and `golangci-lint run ./internal/...` are clean.

**Comment by curry30000** [comment]
@DSFans2014 could you take a look when you have time?

File: internal/manager/manager.go [hunk 1]
```diff
@@ -373,3 +374,10 @@ func (am *AscendManager) IsHamiVnpuCore() bool {
 	}
 	return am.globalConfig.VNPUs.HamiVnpuCore
 }
+
+func (am *AscendManager) DeviceCoreScaling() float64 {
+	if am.nodeConfig != nil && am.nodeConfig.DeviceCoreScaling > 0 {
```
Comment (xrwang8) [inline internal/manager/manager.go]: A negative node override falls through to the global setting here. With global `deviceCoreScaling: 2` and node `deviceCoreScaling: -1`, this returns 2 and the plugin advertises 200. The negative value never reaches `AdvertisedDevcore`, so there's no warning or fallback to 100. I reproduced the result of 200 in a unit test.

The documented fallback for values below 1 should apply to node overrides too. Please distinguish an omitted value from an invalid one, and add a test with a negative node override and a global ratio greater than 1.

**Review by xrwang8 — COMMENTED** [review]
One issue with the per-node fallback.

File: internal/manager/manager.go (hunk 1 above)
Comment (curry30000) [reply internal/manager/manager.go]: Agreed — > 0 read an explicit -1 the same as "not set". Changed to != 0, so only an omitted override (the zero value, matching omitempty) falls back to the global ratio and everything else goes through AdvertisedDevcore, which already warns and drops to the base.

Test added for a -1 node override against a global 2: DeviceCoreScaling() returns -1 and AdvertisedDevcore gives 100, not 200. Covered 0.5 the same way. 192e7b6.

**Comment by curry30000** [comment]
> One issue with the per-node fallback.

@xrwang8 I've updated the PR based on your comments. Please take another look when you have a chance. Thanks.

File: internal/server/server.go [hunk 2]
```diff
@@ -46,8 +46,10 @@ const (
 	VNPUModeAnnotation         = "huawei.com/vnpu-mode"
 	VNPUModeHamiCore           = "hami-core"
 	VNPUNodeSelectorAnnotation = "hami-vnpu-core"
-	// HamiVnpuCoreMaxPercent is the total allocatable core units per device in
-	// soft-slice (hami-vnpu-core) mode, where core requests are percentages.
+	// HamiVnpuCoreMaxPercent is the base allocatable core percentage per device
+	// in soft-slice (hami-vnpu-core) mode, where core requests are percentages.
+	// The capacity advertised to HAMi rises above it when deviceCoreScaling
+	// oversells compute; see internal.AdvertisedDevcore.
```
Comment (DSFans2014) [inline internal/server/server.go]: `HamiVnpuCoreMaxPercent` appears to be unused now. please remove it

Comment (curry30000) [reply internal/server/server.go]: > `HamiVnpuCoreMaxPercent` appears to be unused now. please remove it

OK，Removed. AdvertisedDevcore now owns the percentage budget, so HamiVnpuCoreMaxPercent is no longer referenced.

**Review by DSFans2014 — COMMENTED** [review]
/lgtm

**Comment by curry30000** [comment]
> /lgtm

DCO sign-off was missing on the cleanup commit; added it, no code change. Could you please /lgtm again when you have a minute?

**Comment by curry30000** [comment]
@archlitchi could you take a look when you have time?

**Review by DSFans2014 — COMMENTED** [review]
/lgtm

**Review by Shouren — APPROVED** [review]
/lgtm

**Review by archlitchi — APPROVED** [review]
/lgtm
